### PR TITLE
feat(python): support custom models without forking

### PR DESCRIPTION
**Pull request by hsheth2**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/scripts/custom_package_codegen.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+from setuptools import setup
+
+_package_name = "{package_name}"
+_package_version = "{package_version}"
+
+setup(
+    name=_package_name,
+    version=_package_version,
+    install_requires=[
+        "avro-gen3==0.7.11",
```
Comment (asikowitz) [inline metadata-ingestion/scripts/custom_package_codegen.py:79]: Can we put this into a variable? Perhaps match with the value in setup.py

File: metadata-ingestion/scripts/custom_package_codegen.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        "acryl-datahub",
+    ],
+    entry_points={{
+        "datahub.custom_packages": [
+            f"models={python_package_name}.models.schema_classes",
```
Comment (asikowitz) [inline metadata-ingestion/scripts/custom_package_codegen.py:84]: Is the `f` doing anything here?

**Review by asikowitz — APPROVED** [review]

File: docs/modeling/extending-the-metadata-model.md [hunk 3]
```diff
@@ -251,14 +251,39 @@ Then, run `./gradlew build` from the repository root to rebuild Datahub with acc
...
 to deploy during development. This will allow Datahub to read and write your new entity or extensions to existing entities, along with serving search and graph queries for that entity type.
 
-To emit proposals to ingest from the Datahub CLI tool, first install datahub cli
-locally [following the instructions here](../../metadata-ingestion/developing.md). `./gradlew build` generated the avro
-schemas your local ingestion cli tool uses earlier. After following the developing guide, you should be able to emit
-your new event using the local datahub cli.
+### <a name="step_7"></a>(Optional) Step 7: Use custom models with the Python SDK
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
```
[the remark this answers is not in the document]
Comment (hsheth2) [reply docs/modeling/extending-the-metadata-model.md:257]: the md is actually treated as mdx, so react code like this is valid

File: metadata-ingestion/scripts/avro_codegen.py [hunk 4]
```diff
@@ -343,8 +343,15 @@ class AspectBag(TypedDict, total=False):
     "schemas_path", type=click.Path(exists=True, file_okay=False), required=True
 )
 @click.argument("outdir", type=click.Path(), required=True)
+@click.option("--check-unused-aspects", is_flag=True, default=False)
+@click.option("--enable-custom-loader", is_flag=True, default=True)
 def generate(
```
[the remark this answers is not in the document]
Comment (hsheth2) [reply metadata-ingestion/scripts/avro_codegen.py:348]: this script generates all the files in `src/metadata`

File: metadata-ingestion/scripts/avro_codegen.py [hunk 5]
```diff
@@ -404,6 +414,35 @@ def generate(
...
+from datahub.utilities._custom_package_loader import get_custom_models_package
+
+_custom_package_path = get_custom_models_package()
+
+if TYPE_CHECKING or not _custom_package_path:
+    from ._schema_classes import *
+
+    # Required explicitly because __all__ doesn't include _ prefixed names.
+    from ._schema_classes import _Aspect
+else:
```
[the remark this answers is not in the document]
Comment (hsheth2) [reply metadata-ingestion/scripts/avro_codegen.py:439]: the line is `if TYPE_CHECKING or not _custom_package_path:`

so if there's no custom package, it'll get loaded normally

[the remark this answers is not in the document]
Comment (hsheth2) [reply metadata-ingestion/scripts/avro_codegen.py:439]: will fix the locals -> globals thing

File: metadata-ingestion/scripts/custom_package_codegen.py (hunk 1 above)
Comment (hsheth2) [reply metadata-ingestion/scripts/custom_package_codegen.py]: did it using entrypoints, but yes

File: metadata-ingestion/scripts/custom_package_codegen.py (hunk 2 above)
Comment (hsheth2) [reply metadata-ingestion/scripts/custom_package_codegen.py]: nope - removed

**Pull request merged**